sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -53,6 +53,40 @@ case class Project(projectList: Seq[NamedExpression], child: LogicalPlan) extend
 
   override def validConstraints: Set[Expression] =
     child.constraints.union(getAliasedConstraints(projectList))
+
+  override def sql: String = {
+    if (projectList.exists(expr => expr.find(e => e.isInstanceOf[NonSQLExpression]).isDefined)) {
```

Comment:
This `if` is overly complex and forces a reader to read the expression inside-out (not left to right).

I can't propose anything better than beg for a boolean val with the entire boolean condition on a separate line.

WDYT?